docs/design/2021-03-09-dynamic-privileges.md

```diff
@@ -0,0 +1,282 @@
+# Proposal:
+
+- Author(s):     [morgo](https://github.com/morgo)
+- Last updated:  March 09, 2021
+- Discussion at: N/A
+
+## Table of Contents
+
+* [Introduction](#introduction)
+* [Motivation or Background](#motivation-or-background)
+* [Detailed Design](#detailed-design)
+* [Test Design](#test-design)
+    * [Functional Tests](#functional-tests)
+    * [Scenario Tests](#scenario-tests)
+    * [Compatibility Tests](#compatibility-tests)
+    * [Benchmark Tests](#benchmark-tests)
+* [Impacts & Risks](#impacts--risks)
+* [Investigation & Alternatives](#investigation--alternatives)
+* [Unresolved Questions](#unresolved-questions)
+
+## Introduction
+
+This document was created to discuss the design of Dynamic Privileges. It is intended to be implemented in combination with Security Enhanced Mode, but there no interdependencies between the two features.
+
+## Motivation or Background
+
+MySQL 8.0 introduced the concept of “dynamic privileges” (see [WL#8131](https://dev.mysql.com/worklog/task/?id=8131)). The intention behind this functionality is that plugins can create new named privileges to suit their purposes, such as “Firewall Admin” or “Audit Admin” instead of requiring the `SUPER` privilege, which becomes overloaded and too coarse.
+
+Dynamic privileges are different from static privileges in the following ways:
+
+* The name of the privilege is “DYNAMIC”. A plugin registers it, and the server has no prior knowledge of its existence.
+* Privileges can only be global scoped.
+* Privileges are stored in the table `mysql.global_grants` and not `mysql.user`.
+* The `GRANT OPTION` is stored for each dynamic privilege (versus for the user as a whole).
+
+We have the same requirement for fine grained privileges in TiDB, so it makes sense to adopt a similar implementation of dynamic privileges. This document describes both the implementation of the framework for dynamic privileges and an initial set of dynamic privileges that are required to be implemented.
+
+## Detailed Design
+
+Implementing Dynamic Privileges requires the following work to be completed.
+
+### Persistence
+
+For TiDB, we can use the same table structure as MySQL:
+
+```sql
+CREATE TABLE `global_grants` (
+  `USER` char(32) NOT NULL DEFAULT '',
+  `HOST` char(255) NOT NULL DEFAULT '',
+  `PRIV` char(32) NOT NULL DEFAULT '',
+  `WITH_GRANT_OPTION` enum('N','Y') NOT NULL DEFAULT 'N',
+  PRIMARY KEY (`USER`,`HOST`,`PRIV`)
+);
+```
+
+There is an existing table called “global_priv” which initially looked like it provided similar functionality, except:
```

Comment:
The `global_priv` table does not exist in MySQL. As far as I can tell these columns are stored in `mysql.user` in MySQL.

I think it should be cleaned up to be MySQL compatible so restoring mysqldump can work as expected, but would suggest we handle it outside of this proposal, since for storage this follows MySQL exactly.